### PR TITLE
WIP: GHC 8.2 revamp

### LINKED ISSUE #180: Error with promotion fields for sum types
When I write
```
promote [d|
  data X = X1 {y :: Symbol} | X2 {y :: Symbol}
  z (X1 x) = x
  z (X2 x) = x
  |]
```
I expect that Y and Z are similar. But now definition for Y is wrong:

```
:i Y
type family Y (a0 :: X) :: Symbol
  where [(field0 :: Symbol)] Y ('X2 field0) = field0
:i Z
type family Z (a0 :: X) :: Symbol
  where
    [(x0 :: Symbol)] Z ('X1 x0) = x0
    [(x0 :: Symbol)] Z ('X2 x0) = x0
```
 That is on [stack lts-8.0](https://www.stackage.org/lts-8.0/package/singletons-2.2)

### PR DESCRIPTION
**Pull request by RyanGlScott**
Fixes #180 while I'm in town.

### CODE REVIEW

File: src/Data/Singletons/Promote.hs [hunk 1]
```diff
@@ -552,6 +554,7 @@ promotePat (DBangPa pat) = do
   qReportWarning "Strict pattern converted into regular pattern in promotion"
   (ty, pat') <- promotePat pat
   return (ty, DBangPa pat')
+promotePat (DSigPa _pat _ty) = error "TODO: Handle SigPa"
```
Comment (RyanGlScott) [inline src/Data/Singletons/Promote.hs]: To be honest, I'm not quite sure how one would promote a `DSigPa`. Suggestions welcome.

File: src/Data/Singletons/Single.hs [hunk 2]
```diff
@@ -499,6 +501,7 @@ singPat var_proms patCxt (DTildePa pat) = do
 singPat var_proms patCxt (DBangPa pat) = do
   (pat', ty) <- singPat var_proms patCxt pat
   return (DBangPa pat', ty)
+singPat _var_proms _patCxt (DSigPa _pat _ty) = error "TODO: Handle SigPa"
```
Comment (RyanGlScott) [inline src/Data/Singletons/Single.hs]: Similarly, how do we single `DSigPa`s?

File: src/Data/Singletons/Partition.hs [hunk 3]
```diff
@@ -46,23 +48,35 @@ partitionDecs :: Quasi m => [DDec] -> m PartitionedDecs
...
+         | stock, deriv_name == boundedName
         -> Right <$> mkBoundedInstance ty cons
-         | deriv_name == enumName
+         | stock, deriv_name == enumName
         -> Right <$> mkEnumInstance ty cons
+        where
+          stock = case strat of
+                    Nothing            -> True
+                    Just StockStrategy -> True
+                    Just _             -> False
```
Comment (RyanGlScott) [inline src/Data/Singletons/Partition.hs]: I opted to only choose `deriving` clauses with an explicit strategy when that strategy is `stock`. We could make it more permissive and ignore the strategy, however.

File: src/Data/Singletons/Prelude/List.hs [hunk 4]
```diff
@@ -306,12 +306,18 @@ $(singletonsOnly [d|
...
     where
       perms []     _  = []
       perms (t:ts) is = foldr interleave (perms ts (t:is)) (permutations is)
         where interleave    xs     r = let (_,zs) = interleave' id xs r in zs
+
+              -- This type signature isn't present in the reference
+              -- implementation of permutations in base. However, it is needed
+              -- here, since (at least in GHC 8.2.1) the singletonized version
+              -- will fail to typecheck without it. See #13549 for the full story.
+              interleave' :: ([a] -> b) -> [a] -> [b] -> ([a], [b])
```
Comment (RyanGlScott) [inline src/Data/Singletons/Prelude/List.hs]: It turns out adding one type signature is enough to work around [GHC Trac #13549](https://ghc.haskell.org/trac/ghc/ticket/13549). Hooray!

File: tests/compile-and-dump/Singletons/BadBoundedDeriving.ghc82.template [hunk 5]
```diff
@@ -1,3 +1,6 @@
 
 Singletons/BadBoundedDeriving.hs:0:0: error:
     Can't derive Bounded instance for Foo_0 a_1.
+  |
+6 | $(singletons [d|
+  |   ^^^^^^^^^^^^^^...
```
Comment (RyanGlScott) [inline tests/compile-and-dump/Singletons/BadBoundedDeriving.ghc82.template]: Do you think we should include these error message locations in golden files? They do add some degree of bloat.

**Comment by RyanGlScott** [comment]
Another TODO: the Travis build is failing because `Cabal-2.0` is inclined to build the test suite _before_ the library. Unfortunately, this breaks the assumption that `"../dist/build/autogen/cabal_macros.h"` will exist by the time the test suite is built (see [this Travis error](https://travis-ci.org/goldfirere/singletons/jobs/224782539#L656) for an example).

Perhaps we should just make the test suite depend on `singletons` in `singletons.cabal`? Then again, there's [this Note](https://github.com/goldfirere/singletons/blob/f6516af95f2421f9509abfb992523d430e00b6ad/tests/SingletonsTestSuiteUtils.hs#L121-L124):

```
-- Note [-this-unit-id hack]
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
--
-- We want to avoid installing singletons package before running the
-- testsuite, because in this way we prevent double compilation of the
-- library.
```

I don't think making the test suite depend on `singletons` in `singletons.cabal` would result in double compilation, but then again, I am quite new to the inner machinations of this test suite.

File: src/Data/Singletons/Prelude/List.hs (hunk 4 above)
Comment (goldfirere) [inline src/Data/Singletons/Prelude/List.hs]: See my comment to GHC's #13459. This type signature is indeed necessary: a type signature is required on a local function if its type is inferred differently depending on the presence of `MonoLocalBinds`. This is the case here.

**Comment by goldfirere** [comment]
Somehow just saw your last comment, about Cabal and `-this-unit-id`. It's quite possible that modern Cabals won't cause double-compilation.... or simply that we had configured it wrongly when setting this up. I don't think it would severely break things to add the `singletons` dependency.

**Comment by RyanGlScott** [comment]
OK, adding `singletons` as a test suite dependency seemed to fix that particular problem. But now the tests themselves are failing, apparently because of different unique values. But we enable `-dsuppress-uniques`... hm. I need to stare at this more closely.

**Comment by RyanGlScott** [comment]
Ah, here's a hunch. In the test suite, we do [normalization with `sed`](https://github.com/goldfirere/singletons/blob/023513fe2ee63391095540f575873064bbe3ab03/tests/SingletonsTestSuiteUtils.hs#L199-L218):

```
-- Note [Normalization with sed]
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
--
-- Output file is normalized with sed. Line numbers generated in splices:
--
--   Foo:(40,3)-(42,4)
--   Foo.hs:7:3:
--   Equals_1235967303
--
-- are turned into:
--
--   Foo:(0,0)-(0,0)
--   Foo.hs:0:0:
--   Equals_0123456789
```

In particular, we normalize [this pattern](https://github.com/goldfirere/singletons/blob/023513fe2ee63391095540f575873064bbe3ab03/tests/SingletonsTestSuiteUtils.hs#L230):

```haskell
"-e", "'s/[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]/0123456789/g'"
```

This works assuming that whenever you have something like `Equals_1235967303`, it will always have exactly 10 numbers following it. But in GHC 8.2, uniques appear to have 19 numbers (perhaps due to [this commit](http://git.haskell.org/ghc.git/commit/0d213c18b6962bb65e2b3035a258dd3f5bf454dd)?). Maybe adding a pattern for 19 numbers will do the trick.

**Comment by RyanGlScott** [comment]
Luckily, adding more patterns for `sed` was precisely the secret sauce needed to cook up a green Travis build.

**Comment by RyanGlScott** [comment]
Should this be closed now that 637fb4d0435220f740bc658973f7a05a499dcee6 has been committed?

**Comment by goldfirere** [comment]
Don't close quite yet. I did merge (it passes the testsuite) but I haven't reviewed carefully yet. Will do so soon.

**Comment by goldfirere** [comment]
This has been merged.

**Pull request closed without merging**